src/main/groovy/com/google/protobuf/gradle/GenerateProtoTask.groovy

```diff
@@ -0,0 +1,311 @@
+      if (dir instanceof File) {
+          includeDirs.add(dir)
+      } else {
+          includeDirs.add(project.file(dir))
+      }
+  }
+
+  /**
+   * The container of command-line options for a protoc plugin or a built-in output.
+   */
+  public static class PluginOptions implements Named {
+    private final ArrayList<String> options = new ArrayList<String>()
+    private final String name
+
+    public PluginOptions(String name) {
+      this.name = name
+    }
+
+    /**
+     * Adds a plugin option.
+     */
+    public PluginOptions option(String option) {
+      options.add(option)
+      return this
+    }
+
+    public List<String> getOptions() {
+      return options
+    }
+
+    /**
+     * Returns the name of the plugin or builtin.
+     */
+    @Override
+    public String getName() {
+      return name
+    }
+  }
+
+  //===========================================================================
+  //    protoc invocation logic
+  //===========================================================================
+
+  // protoc allows you to prefix comma-delimited options to the path in
+  // the --*_out flags, e.g.,
+  // - Without options: --java_out=/path/to/output
+  // - With options: --java_out=option1,option2:/path/to/output
+  // This method generates the prefix out of the given options.
+  def String makeOptionsPrefix(List<String> options) {
+    StringBuilder prefix = new StringBuilder()
+    if (!options.isEmpty()) {
+      options.each { option ->
+        if (prefix.length() > 0) {
+          prefix.append(',')
+        }
+        prefix.append(option)
+      }
+      prefix.append(':')
+    }
+    return prefix.toString()
+  }
+
+  String getOutputDir(PluginOptions plugin) {
+    return "${outputBaseDir}/${plugin.name}"
+  }
+
+  Collection<File> getAllOutputDirs() {
+    ImmutableList.Builder<File> dirs = ImmutableList.builder()
+    builtins.each { builtin ->
+      dirs.add(new File(getOutputDir(builtin)))
+    }
+    plugins.each { plugin ->
+      dirs.add(new File(getOutputDir(plugin)))
+    }
+    return dirs.build()
+  }
+
+  @TaskAction
+  def compile() {
+    Preconditions.checkState(!initializing, 'doneInitializing() has not been called')
+
+    // Only at this point all outputs are finalized, then we can register the
+    // generated source dirs to java compilation.
+
+    // NOTE(zhangkun83): does this mean we can actually allow per-plugin output
+    // dir reconfiguraiton? Not actually -- the basedir of all outputs must be
+    // registered to task.outputs before Gradle starts to execute tasks, so
+    // that the task can be correctly skipped.  We may only allow plugins to
+    // output to a customized sub dir, but not to a random place, esp. out of
+    // the pre-designated basedir. I doubt the usefulness of this
+    // configurability given such restriction.
+    if (Utils.isAndroidProject(project)) {
+      variant.registerJavaGeneratingTask(this, getAllOutputDirs())
```

Comment:
I have moved the dependency setup and output wiring to right after the execution of ``generateProtoTasks`` closures (in ``ProtobufPlugin.apply()`` by calling ``linkGenerateProtoTasksToJavaCompile``), basically the alternative locations I have tried won't work well:
 1. At the beginning of ``GenerateProtoTask.compile()``. In incremental builds, the task may be skipped when proto files are not changed, but we still want to do the output wiring, and bummer!                                                                                                                            
 2. In ``project.gradle.taskGraph.beforeTask`` closure, which is run in the task execution phase. Android plugin doesn't allow calling ``variant.registerJavaGeneratingTask()`` in this phase.
                                                                                    
Caveat: if the user configures the ``generateProto`` tasks in their own ``project.afterEvaluate`` closure, e.g., adding a new builtin/plugin, it won't be wired to the javaCompile inputs. So, user has to configure tasks in ``protobuf.generateProtoTasks``.  I believe this constraint is acceptable.